validation: add reusable coins view for ConnectBlock

andrewtoth opened the pull request:
This is the first commit of #31132, which can be merged as an independent change. It has a small benefit on its own, but will help in moving the parent PR forward.

Add a `Reset()` method to `CCoinsViewCache` that clears `cacheCoins`, `cachedCoinsUsage`, and `hashBlock` without flushing to the `base` view. This allows efficiently reusing a cache instance across multiple blocks.

Add `CCoinsViewCache::CreateResetGuard` method to return a `CCoinsViewCache::ResetGuard`. The `ResetGuard` automatically calls `Reset()` on destruction. This RAII pattern ensures the cache is always properly reset between blocks.

Add `m_connect_block_view` as a persistent `CCoinsViewCache` for `ConnectBlock`, avoiding repeated memory allocations.

l0rinc commented on src/coins.cpp:
```diff
@@ -275,6 +275,13 @@ bool CCoinsViewCache::Sync()
+{
+    cacheCoins.clear();
+    cachedCoinsUsage = 0;
+    hashBlock.SetNull();
```
This resets `hashBlock`, but `Flush()`/`Sync()` pass the raw `hashBlock` into `base->BatchWrite()`:
* https://github.com/bitcoin/bitcoin/blob/adbb4b320834516dc97fa33aaa853a45cbabca67/src/coins.cpp#L252-L254
* https://github.com/bitcoin/bitcoin/blob/adbb4b320834516dc97fa33aaa853a45cbabca67/src/coins.cpp#L265-L268
This means that the behavior of these methods depends on whether the `GetBestBlock` was called before these methods (since the "getter" would set the field after which accessing it directly would yield a different result):
https://github.com/bitcoin/bitcoin/blob/adbb4b320834516dc97fa33aaa853a45cbabca67/src/coins.cpp#L176-L180
This isn't just theoretical, we have calls to that getter only for its side-effect:
https://github.com/bitcoin/bitcoin/blob/adbb4b320834516dc97fa33aaa853a45cbabca67/src/validation.cpp#L874-L876
I understand that some of this wasn't introduced in this commit, but the presence of a public `Reset` makes this potential misuse more obvious.

----


Nit: `BatchWrite()` ends with `hashBlock = hashBlockIn`, while there’s also a `SetBestBlock()` setter. Consider using the setter consistently (or dropping it) so best-block updates go through one path (helps with debugging and consistency).

andrewtoth replied:
Done, I use the setter instead of setting directly.

l0rinc replied:
We will need to deal with the lazy `GetBestBlock` eventually - I agree it's not a problem here, can be done in a follow-up.

l0rinc commented on src/validation.cpp:
```diff
@@ -3101,7 +3102,7 @@ bool Chainstate::ConnectTip(
              Ticks<MillisecondsDouble>(time_2 - time_1));
     {
-        CCoinsViewCache view(&CoinsTip());
+        auto& view{*m_coins_views->m_connect_block_view};
```
This PR reuses the cache in `ConnectTip()` only, but we still construct short-lived `CCoinsViewCache` instances in `DisconnectTip()` and `TestBlockValidity()`:
* https://github.com/bitcoin/bitcoin/blob/adbb4b320834516dc97fa33aaa853a45cbabca67/src/validation.cpp#L2981
* https://github.com/bitcoin/bitcoin/blob/adbb4b320834516dc97fa33aaa853a45cbabca67/src/validation.cpp#L4700

If the goal is avoiding repeated allocations, should we also consider reusing the persistent view there (maybe as a follow-up to keep this PR small)?

l0rinc reviewed (changes requested):
Concept ACK, carving this out is a small step in speedup up IBD and simplifying the review for the heavier https://github.com/bitcoin/bitcoin/pull/31132

I want to make sure we make the code safer than before, so I'm suggesting we cleanup up the spread-around state changes a bit since it's hard to reason about it otherwise - especially trying to imagine a multi-threaded context where lazy inits and resets and direct setters and field accesses are potentially competing.

andrewtoth commented on src/validation.cpp:
```diff
@@ -3119,8 +3121,9 @@ bool Chainstate::ConnectTip(
                  Ticks<SecondsDouble>(m_chainman.time_connect_total),
                  Ticks<MillisecondsDouble>(m_chainman.time_connect_total) / m_chainman.num_blocks_total);
-        bool flushed = view.Flush(/*will_reuse_cache=*/false); // local CCoinsViewCache goes out of scope
+        bool flushed = view.Flush(/*will_reuse_cache=*/false); // No need to reallocate since it only has capacity for 1 block
```
Done in a separate commit.

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -85,6 +85,11 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsView& backend
+            [&] {
+                coins_view_cache.Reset();
+                // Set best block hash to non-null to satisfy the assertion in CCoinsViewDB::BatchWrite().
+                if (is_db) coins_view_cache.SetBestBlock(uint256::ONE);
```
what if we randomized this, too?

l0rinc reviewed (approved):
Redid the change locally, ended up with slightly different setup, but your solution might handle the lifecycle slightly better.
I ran a `reindex-chainstate` benchmark, the difference is within noise 👍. 
Left some comments, but I'm fine with merging as is:
ACK b46b61bdcefbaa649fc73b8f8936bc80468de763

andrewtoth committed:
refactor: rename will_reuse_cache to reallocate_cache

More accurately reflects the purpose of the parameter, since
we will keep reusing the cache but don't want to reallocate it.

andrewtoth committed:
coins: add Reset on CCoinsViewCache

Add a Reset() method to CCoinsViewCache that clears cacheCoins,
cachedCoinsUsage, and hashBlock without flushing to the base view.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>
Co-authored-by: sedited <seb.kung@gmail.com>

andrewtoth committed:
coins: introduce CCoinsViewCache::ResetGuard

CCoinsViewCache::CreateResetGuard returns a guard that calls
Reset on the cache when the guard goes out of scope.
This RAII pattern ensures the cache is always properly reset
when it leaves current scope.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>
Co-authored-by: sedited <seb.kung@gmail.com>

andrewtoth committed:
validation: reuse same CCoinsViewCache for every ConnectBlock call

Add m_connect_block_view to ChainState's CoinsViews.
Call CreateResetGuard inside ConnectTip to ensure the view
is Reset after each block, avoiding repeated memory allocations.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

andrewtoth committed:
coins: use hashBlock setter internally for CCoinsViewCache methods

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

l0rinc reviewed (commented):
I want to see if there's a simpler way to do automatic cleanup via smart pointers, pushing my comments in the meantime.

sipa reviewed (commented):
Code review ACK 020a624a6995330828fb6a0f762621c6ff498d69. Only nits.

andrewtoth commented:
Thank you @sipa. I've addressed your nits.

`git diff 020a624a6995330828fb6a0f762621c6ff498d69..b52435c25b08850c71d299ab63e4f2fd305fa0ef`

sedited reviewed (commented):
Concept ACK

sedited reviewed (approved):
ACK 3e0fd0e4ddd894f0e7db1772f10ceaa1dddfb951

Would be good to get another pair of eyes.

achow101 commented:
ACK 3e0fd0e4ddd894f0e7db1772f10ceaa1dddfb951
